File: sdk-python/obsvr/ssrf.py
"""SSRF guard for outbound backend URLs (ADR-4 external policy backend).

Twin of sdk-typescript/src/utils/ssrf.ts. An external policy backend URL is UNTRUSTED
input (it may arrive from server-pushed config or a multi-tenant dashboard), so
a naive request would be a server-side request forgery primitive. This module
blocks non-http(s) schemes and private / loopback / link-local / metadata
addresses, RESOLVING the hostname before connecting so a public-looking name
that resolves to a private IP is still refused.

Two tiers:
  - ALWAYS blocked (even with allow_private_network): cloud metadata + link-local.
  - Blocked BY DEFAULT, allowed with allow_private_network=True: loopback +
    RFC1918 / ULA / CGNAT private ranges (a legit OPA/Cedar sidecar is often on
    localhost or a private host).

Stdlib only (``ipaddress`` + ``socket``) to keep the core dependency-free.
"""

import ipaddress
import re
import socket
from typing import Callable, List, Optional
from urllib.parse import urlsplit


class SsrfError(Exception):
    """Raised when a backend URL fails the SSRF guard."""


#: An IPv4 address written entirely in numeric parts, in any radix, with the
#: 1-, 2- and 3-part short forms allowed: ``2852039166``, ``0251.0376.0251.0376``,
#: ``0xa9.0xfe.0xa9.0xfe``, ``0xa9fea9fe``. Every part must be a numeric literal,
#: so a real hostname (any part carrying a non-hex letter, or a label that is not
#: a number at all) never matches and is left to DNS resolution.
_IPV4_NUMERIC = re.compile(
    r"^(?:0[xX][0-9a-fA-F]+|0[0-7]*|[1-9][0-9]*)"
    r"(?:\.(?:0[xX][0-9a-fA-F]+|0[0-7]*|[1-9][0-9]*)){0,3}\.?$"
)


def _parse_ipv4_numeric(host: str) -> Optional[ipaddress.IPv4Address]:
    """A NON-CANONICAL IPv4 literal, normalized to its dotted-quad address.

    ``ipaddress.ip_address`` accepts dotted-quad only, so every other spelling
    of the same address fell through this module as though it were a hostname
    and skipped the range check entirely. Measured before this existed: with
    ``169.254.169.254`` refused, ``https://2852039166/``,
    ``https://0251.0376.0251.0376/``, ``https://0xa9.0xfe.0xa9.0xfe/`` and
    ``https://0xa9fea9fe/`` were all ACCEPTED -- four spellings of the cloud
    metadata endpoint past a guard whose whole purpose is that endpoint, plus
    ``https://2130706433/`` for loopback. The resolver reads all of them as the
    address they encode, so the connection went where the guard said it would
    not.

    This is the same defect ``_unwrap`` documents for IPv6: one spelling of an
    address refused while another spelling of the SAME address is admitted. The
    arithmetic is delegated to ``inet_aton``, which is what the platform
    resolver itself uses, after the pattern above has established that every
    part really is a numeric literal -- ``inet_aton`` alone is lenient about
    trailing content on some platforms.

    Twin: the TypeScript guard gets this for free from the WHATWG URL parser,
    which normalizes these spellings before ``hostname`` is ever read; it
    already refused all five.
    """
    if not _IPV4_NUMERIC.match(host):
        return None
    try:
        return ipaddress.IPv4Address(socket.inet_aton(host.rstrip(".")))
    except (OSError, ipaddress.AddressValueError):
        return None


def _parse_ip(ip: str) -> Optional[ipaddress._BaseAddress]:
    host = ip.strip().strip("[]")
    try:
        return ipaddress.ip_address(host)
    except ValueError:
        # Not dotted-quad or a v6 literal - it may still be an IPv4 address in
        # another radix, which every resolver accepts and this guard must too.
        return _parse_ipv4_numeric(host)


def _unwrap(addr: ipaddress._BaseAddress) -> ipaddress._BaseAddress:
    """Unwrap an IPv6 address that CARRIES an IPv4 address to that IPv4.

    This used to cover exactly one of the four ways an IPv6 literal can carry a
    v4 address -- ``ipaddress.ipv4_mapped``, i.e. ``::ffff:a.b.c.d``. Measured:
    with only that form folded, ``http://[::169.254.169.254]/`` reached the
    network through the external-policy backend and was ACCEPTED by ``init()``
    as a Presidio URL, while the ``::ffff:`` spelling of the SAME address was
    refused -- so "always blocked, no opt-out" over the cloud-metadata endpoint
    held for one spelling of it.

    The four forms, all of which route to the same host:
      - IPv4-MAPPED      ``::ffff:a.b.c.d``           (stdlib ``ipv4_mapped``)
      - IPv4-COMPATIBLE  ``::a.b.c.d``                deprecated, still routable
      - NAT64            ``64:ff9b::a.b.c.d``         RFC 6052 / 8215
      - 6to4             ``2002:HHHH:HHHH::``         RFC 3056 -- the v4 address
                                                      is the two groups after
                                                      the prefix

    Twin: ``mappedIpv4`` in sdk-typescript/src/utils/ssrf.ts.
    """
    if not isinstance(addr, ipaddress.IPv6Address):
        return addr
    if addr.ipv4_mapped is not None:
        return addr.ipv4_mapped
    packed = addr.packed
    # IPv4-compatible: 96 zero bits then the v4 address. `::` and `::1` are
    # loopback/unspecified and are classified on their own merits, not folded.
    if packed[:12] == b"\x00" * 12 and packed[12:] not in (b"\x00\x00\x00\x00", b"\x00\x00\x00\x01"):
        return ipaddress.IPv4Address(packed[12:])
    # NAT64 well-known prefix 64:ff9b::/96.
    if packed[:12] == b"\x00\x64\xff\x9b" + b"\x00" * 8:
        return ipaddress.IPv4Address(packed[12:])
    # 6to4: 2002:V4V4:V4V4::/48.
    if packed[:2] == b"\x20\x02":
        return ipaddress.IPv4Address(packed[2:6])
    return addr


def is_ip_literal(host: str) -> bool:
    """True when ``host`` is an IPv4/IPv6 literal (brackets already stripped)."""
    return _parse_ip(host) is not None


def is_always_blocked_ip(ip: str) -> bool:
    """Cloud-metadata + link-local addresses. ALWAYS blocked — no opt-out.
    Covers 169.254.0.0/16 (incl. 169.254.169.254), IPv6 link-local fe80::/10,
    and the AWS IPv6 metadata address fd00:ec2::254."""
    addr = _parse_ip(ip)
    if addr is None:
        return False
    addr = _unwrap(addr)
    if addr.is_link_local:  # 169.254/16 and fe80::/10
        return True
    if isinstance(addr, ipaddress.IPv6Address) and addr == ipaddress.IPv6Address("fd00:ec2::254"):
        return True
    return False


def is_private_or_reserved_ip(ip: str) -> bool:
    """Private / loopback / reserved (blocked by default; allowed with
    allow_private_network). Uses the stdlib classifiers plus an explicit
    100.64.0.0/10 (CGNAT) check for parity with the TS twin."""
    addr = _parse_ip(ip)
    if addr is None:
        return False
    addr = _unwrap(addr)
    if (
        addr.is_private
        or addr.is_loopback
        or addr.is_link_local
        or addr.is_reserved
        or addr.is_multicast
        or addr.is_unspecified
    ):
        return True
    # CGNAT 100.64.0.0/10 — not flagged is_private on every Python version.
    if isinstance(addr, ipaddress.IPv4Address) and addr in ipaddress.ip_network("100.64.0.0/10"):
        return True
    return False


def _assert_ip_allowed(ip: str, allow_private_network: bool) -> None:
    if is_always_blocked_ip(ip):
        raise SsrfError(f"backend url resolves to a blocked metadata/link-local address: {ip}")
    if not allow_private_network and is_private_or_reserved_ip(ip):
        raise SsrfError(
            f"backend url resolves to a blocked private/reserved address: {ip} "
            "(set allow_private_network to permit a sidecar/private-network backend)"
        )


def assert_backend_url_static(url: str, allow_private_network: bool = False):
    """STATIC guard (no DNS): validate scheme and, for a literal-IP host, its
    range. Called at init() so a clearly-unsafe URL fails fast."""
    try:
        parts = urlsplit(url)
    except ValueError as e:
        raise SsrfError(f"invalid backend url: {url!r}") from e
    if parts.scheme not in ("http", "https"):
        raise SsrfError(f'backend url scheme must be http(s), got "{parts.scheme}"')
    host = parts.hostname or ""
    if host == "":
        raise SsrfError(f"backend url has no host: {url!r}")
    if is_ip_literal(host):
        _assert_ip_allowed(host, allow_private_network)
    return parts


Resolver = Callable[[str], List[str]]


def _default_resolver(host: str) -> List[str]:
    infos = socket.getaddrinfo(host, None)
    return [info[4][0] for info in infos]


def assert_backend_url_allowed(
    url: str,
    allow_private_network: bool = False,
    resolver: Optional[Resolver] = None,
) -> None:
    """FULL guard: static checks plus DNS resolution of a hostname host,
    asserting EVERY resolved address is permitted (resolve-before-connect). A
    public name that resolves to a private IP is refused. Literal-IP hosts skip
    DNS. Residual TOCTOU on hostnames is documented in the TS twin."""
    parts = assert_backend_url_static(url, allow_private_network)
    host = parts.hostname or ""
    if is_ip_literal(host):
        return
    resolve = resolver or _default_resolver
    addrs = resolve(host)
    if not addrs:
        raise SsrfError(f"backend url host did not resolve: {host}")
    for addr in addrs:
        _assert_ip_allowed(addr, allow_private_network)

File: sdk-typescript/src/utils/ssrf.ts
/**
 * SSRF guard for outbound backend URLs (ADR-4 external policy backend).
 *
 * An external policy backend URL is treated as UNTRUSTED input: it may arrive
 * from server-pushed config or a multi-tenant dashboard, so a naive fetch would
 * be a server-side request forgery primitive — an attacker could point it at the
 * cloud metadata endpoint (169.254.169.254) or an internal service. This module
 * blocks non-http(s) schemes and private / loopback / link-local / metadata
 * addresses, RESOLVING the hostname before connecting so a public-looking name
 * that resolves to a private IP is still refused.
 *
 * Two tiers of blocking:
 *  - ALWAYS blocked (even with allowPrivateNetwork): cloud metadata + link-local.
 *    These are never a legitimate policy backend and are the crown-jewel SSRF
 *    target, so no opt-out exists.
 *  - Blocked BY DEFAULT, allowed with allowPrivateNetwork=true: loopback and
 *    RFC1918 / ULA / CGNAT private ranges. A legitimate OPA/Cedar deployment is
 *    frequently a sidecar on localhost or a private-network host, so this
 *    deliberate, documented opt-in keeps the feature usable without weakening
 *    the metadata protection.
 *
 * Kept dependency-free (Node stdlib dns only) to match the SDK's zero-runtime-dep
 * posture. Twin: sdk-python/obsvr/ssrf.py — the two must classify the same
 * private/reserved IP ranges identically.
 *
 * @packageDocumentation
 */

import { lookup } from 'node:dns/promises';

/** Raised when a backend URL fails the SSRF guard. */
export class SsrfError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'SsrfError';
  }
}

export interface SsrfOptions {
  /** Allow loopback + RFC1918/ULA/CGNAT private ranges (never metadata/link-local). */
  allowPrivateNetwork?: boolean;
}

/** Parse an IPv4 dotted-quad into 4 octets, or null if not an IPv4 literal. */
function parseIpv4(ip: string): [number, number, number, number] | null {
  const m = /^(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})$/.exec(ip);
  if (!m) return null;
  const octets = [Number(m[1]), Number(m[2]), Number(m[3]), Number(m[4])];
  if (octets.some((o) => o > 255)) return null;
  return octets as [number, number, number, number];
}

/** Render a 32-bit value carried in two hex groups as a dotted quad. */
function hexPairToIpv4(hi: string, lo: string): string {
  const h = parseInt(hi, 16);
  const l = parseInt(lo, 16);
  return `${(h >> 8) & 0xff}.${h & 0xff}.${(l >> 8) & 0xff}.${l & 0xff}`;
}

/**
 * If `ip` is an IPv6 address that CARRIES an IPv4 address, return that IPv4 as
 * a dotted quad; otherwise null.
 *
 * This is load-bearing for the SSRF guard, and it used to cover exactly one of
 * the four ways an IPv6 literal can carry a v4 address. Measured: with only the
 * `::ffff:` form folded, `http://[::169.254.169.254]/` reached `fetch` through
 * the external-policy backend and was ACCEPTED by `init()` as a Presidio URL,
 * while the `::ffff:` spelling of the same address was refused — so the
 * "always blocked, no opt-out" claim over the cloud-metadata endpoint held for
 * one spelling of it.
 *
 * The four forms, all of which route to the same host:
 *   - IPv4-MAPPED      `::ffff:a.b.c.d` / `::ffff:HHHH:HHHH` / `0:0:0:0:0:ffff:…`
 *   - IPv4-COMPATIBLE  `::a.b.c.d` / `::HHHH:HHHH` — deprecated, still routable
 *   - NAT64            `64:ff9b::a.b.c.d` and `64:ff9b:1::…` (RFC 6052/8215)
 *   - 6to4             `2002:HHHH:HHHH::…` (RFC 3056) — the v4 address is the
 *                      two groups after the prefix
 *
 * Node's WHATWG URL parser normalizes a dotted tail to the hex form, so both
 * spellings of every one of these must be recognized.
 */
function mappedIpv4(ip: string): string | null {
  const norm = ip.toLowerCase().replace(/^\[|\]$/g, "");
  const DOT = "(\\d{1,3}\\.\\d{1,3}\\.\\d{1,3}\\.\\d{1,3})";
  const HEX = "([0-9a-f]{1,4}):([0-9a-f]{1,4})";
  const tail = `(?:${DOT}|${HEX})`;

  // IPv4-mapped: leading zero groups compressed (`::`) or explicit.
  let m = new RegExp(`^(?:::|(?:0:){1,5})ffff:${tail}$`).exec(norm);
  // IPv4-compatible (deprecated, and the form the guard missed).
  if (!m) m = new RegExp(`^(?:::|(?:0:){1,6})${tail}$`).exec(norm);
  // NAT64 well-known prefix, both the /96 and the /48-with-suffix spellings.
  if (!m) m = new RegExp(`^64:ff9b(?::0)*::${tail}$`).exec(norm);
  if (!m) m = new RegExp(`^64:ff9b:1:(?:0:){0,3}:?${tail}$`).exec(norm);
  // 6to4: the two groups immediately after the prefix ARE the v4 address.
  // Anything after them is the site's own subnetting and does not change the
  // host the packet reaches, so the match deliberately ignores the remainder.
  if (!m) {
    const six = /^2002:([0-9a-f]{1,4}):([0-9a-f]{1,4})(?::|$)/.exec(norm);
    if (six) return hexPairToIpv4(six[1], six[2]);
  }

  if (!m) return null;
  if (m[1]) {
    const octets = parseIpv4(m[1]);
    return octets ? m[1] : null;
  }
  if (m[2] && m[3]) return hexPairToIpv4(m[2], m[3]);
  return null;
}

/** True for an IPv4 or IPv6 literal (with brackets already stripped). */
export function isIpLiteral(host: string): boolean {
  if (parseIpv4(host)) return true;
  // Any colon means an IPv6 literal (possibly with a ::ffff: mapped tail).
  return host.includes(':');
}

/**
 * Cloud-metadata + link-local addresses. ALWAYS blocked — no opt-out.
 * Covers 169.254.0.0/16 (incl. 169.254.169.254), IPv6 link-local fe80::/10,
 * and the AWS IPv6 metadata address fd00:ec2::254.
 */
export function isAlwaysBlockedIp(ip: string): boolean {
  // Fold any IPv4-mapped IPv6 form to its IPv4 first (hex or dotted), so a
  // mapped metadata/link-local address cannot slip past the v4 range checks.
  const effective = mappedIpv4(ip) ?? ip;
  const v4 = parseIpv4(effective);
  if (v4) {
    return v4[0] === 169 && v4[1] === 254; // 169.254/16 link-local (metadata)
  }
  const norm = effective.toLowerCase().replace(/^\[|\]$/g, '');
  if (/^fe[89ab]/.test(norm)) return true; // fe80::/10 link-local
  if (norm === 'fd00:ec2::254') return true; // AWS IPv6 metadata
  return false;
}

/**
 * Private / loopback / reserved addresses (blocked by default; allowed with
 * allowPrivateNetwork). Includes 0/8, 10/8, 127/8, 169.254/16, 172.16/12,
 * 192.168/16, 100.64/10 (CGNAT), multicast/reserved (>=224), IPv6 ::/::1,
 * fc00::/7 (ULA), fe80::/10 (link-local), and ::ffff: mapped IPv4.
 */
export function isPrivateOrReservedIp(ip: string): boolean {
  const effective = mappedIpv4(ip) ?? ip;
  const v4 = parseIpv4(effective);
  if (v4) {
    const [a, b] = v4;
    if (a === 0) return true; // 0.0.0.0/8 "this network"
    if (a === 10) return true; // 10/8
    if (a === 127) return true; // loopback
    if (a === 169 && b === 254) return true; // link-local
    if (a === 172 && b >= 16 && b <= 31) return true; // 172.16/12
    if (a === 192 && b === 168) return true; // 192.168/16
    if (a === 100 && b >= 64 && b <= 127) return true; // 100.64/10 CGNAT
    // Reserved documentation / benchmarking ranges (kept in lockstep with the
    // Python twin's ipaddress classification so both SDKs refuse the same set):
    if (a === 192 && b === 0 && (v4[2] === 0 || v4[2] === 2)) return true; // 192.0.0/24 + 192.0.2/24 (TEST-NET-1)
    if (a === 198 && b === 51 && v4[2] === 100) return true; // 198.51.100/24 (TEST-NET-2)
    if (a === 203 && b === 0 && v4[2] === 113) return true; // 203.0.113/24 (TEST-NET-3)
    if (a === 198 && (b === 18 || b === 19)) return true; // 198.18/15 benchmarking
    if (a >= 224) return true; // 224/4 multicast + 240/4 reserved
    return false;
  }
  const norm = effective.toLowerCase().replace(/^\[|\]$/g, '');
  if (norm === '::1' || norm === '::') return true; // loopback / unspecified
  if (/^f[cd]/.test(norm)) return true; // fc00::/7 unique-local
  if (/^fe[89ab]/.test(norm)) return true; // fe80::/10 link-local
  return false;
}

/** Assert one resolved/literal address is permitted, or throw SsrfError. */
export function assertIpAllowed(ip: string, opts: SsrfOptions = {}): void {
  if (isAlwaysBlockedIp(ip)) {
    throw new SsrfError(`backend url resolves to a blocked metadata/link-local address: ${ip}`);
  }
  if (!opts.allowPrivateNetwork && isPrivateOrReservedIp(ip)) {
    throw new SsrfError(
      `backend url resolves to a blocked private/reserved address: ${ip} ` +
        `(set allowPrivateNetwork to permit a sidecar/private-network backend)`,
    );
  }
}

/**
 * STATIC guard (no DNS): validate scheme and, for a literal-IP host, its range.
 * Called at init() so a clearly-unsafe URL fails fast. Hostnames pass this
 * stage and are checked dynamically by {@link assertBackendUrlAllowed}.
 */
export function assertBackendUrlStatic(url: string, opts: SsrfOptions = {}): URL {
  let u: URL;
  try {
    u = new URL(url);
  } catch {
    throw new SsrfError(`invalid backend url: ${String(url)}`);
  }
  if (u.protocol !== 'http:' && u.protocol !== 'https:') {
    throw new SsrfError(`backend url scheme must be http(s), got "${u.protocol}"`);
  }
  const host = u.hostname.replace(/^\[|\]$/g, '');
  if (isIpLiteral(host)) {
    assertIpAllowed(host, opts);
  }
  return u;
}

/** Resolve a hostname to all A/AAAA addresses. Injectable for tests. */
export type Resolver = (host: string) => Promise<string[]>;

const defaultResolver: Resolver = async (host: string): Promise<string[]> => {
  const results = await lookup(host, { all: true });
  return results.map((r) => r.address);
};

/**
 * FULL guard: static checks plus DNS resolution of a hostname host, asserting
 * EVERY resolved address is permitted (resolve-before-connect). A public name
 * that resolves to a private IP is refused. Literal-IP hosts skip DNS.
 *
 * Residual TOCTOU: the subsequent fetch re-resolves the name, so a name that
 * flips to a private IP between this check and the connect could still be hit.
 * Literal-IP backends have no such gap; for hostnames this narrows, not closes,
 * the window. Callers fail closed on any thrown SsrfError.
 */
export async function assertBackendUrlAllowed(
  url: string,
  opts: SsrfOptions = {},
  resolver: Resolver = defaultResolver,
): Promise<void> {
  const u = assertBackendUrlStatic(url, opts);
  const host = u.hostname.replace(/^\[|\]$/g, '');
  if (isIpLiteral(host)) return; // already asserted in the static pass
  const addrs = await resolver(host);
  if (!addrs.length) {
    throw new SsrfError(`backend url host did not resolve: ${host}`);
  }
  for (const addr of addrs) assertIpAllowed(addr, opts);
}
